### ISSUE TITLE
Chapter titles chopped after 50 characters

### ISSUE THREAD

**Issue by Kazaa3** [rmcrackan/Libation#183]
I have some audiobooks that are 20- to 40-hour collections by one author that are impractical to use, but cheap to buy. Which is why I split them up into individual books. The first step I take is to save the audiobook in separated chapters. This is possible with Libation (set Split my books into multiple files by chapter checkbox). 

Unfortunately all chapter titles are chopped off after 50 characters. I have shortened the templet to <ch# 0> - <ch title>, but this does not change anything (and choosing a folder that has a short save path). There seems to be a fixed string limit of 50 characters.

Would it be possible to at least store the full unabridged chapter title in <title> tag, or to use an additional parameter to specify to how many digits the tag is shortened (e.g.: <ch title, 256>) to shorten to 256 characters.

Best regards and thanks for this great piece of software.

**Comment by rmcrackan** [comment]
@Mbucari : When splitting by chapter, does AAXClean store full chapter names in file metatags?

**Comment by Mbucari** [comment]
Yes. ChapterInfo.Title is written into the chapter tags.

**Comment by Mbucari** [comment]
@rmcrackan I actually got around to testing this, and I can indeed writer custom chapter titles longer than 50 characters.  I just tested with a 150 character string and it writes no problem.

**Comment by rmcrackan** [comment]
@Mbucari this one's on me. The original poster means 50 char limit for file name. This is true. Especially with custom file names, the logic for handing file and folder names is really complex. To have a prayer of getting it right, I limit the file name to 50.

**Comment by rmcrackan** [comment]
@Mbucari I thought I saw something in this weekend's big batches about 50 character stuff. Was that about #183 ? Or am I just remembering wrong?

**Comment by Mbucari** [comment]
I totally missed this, but no, I haven't touched anything relating to file naming or trimming chapter titles.

**Comment by Mbucari** [comment]
@rmcrackan
What exactly is the issue here? Is it that we are running up against the maximum file name length in windows, in which case nothing can be done and this issue should be closed? Or is it that these file names could be longer, but something within libation is preventing it?

**Comment by rmcrackan** [comment]
The issue is that handling the resulting file names -- which can get ridiculously long, Esp when many templating fields are used. The easy fix is seen in `\Source\FileManager\FileNamingTemplate.cs` ... `ParameterMaxSize = 50`. If we get rid of ParameterMaxSize and its logic then we can let the chips fall where they may. My fear is that it might lead to even worse paths once it trickles down to FileManager.FileUtility.GetValidFilename . An "easy" fix might be to do exactly this and see how many weird bug reports come in. Maybe I over-thought the problem and users won't actually be affected.

**Comment by Mbucari** [comment]
I looked into this briefly a couple of weeks ago. I don't remember exactly where it wasn in the code base, but I do remember seeing somewhere where you check the file name length against Windows max file name length (260) and check the full path length against the Windows max path length 247. If you're using that logic to truncate file names that exceed the system's max supported length, then you shouldn't need to truncate anywhere else in the file naming process.

But I might be useful to add support for extended length pads using the `\\?\` path prefix. Using that prefix gives you a max path length of 32,767 chars, but you still have a max file name length of 260 chars.

**Comment by rmcrackan** [comment]
> `\\?\`

Cool. That'd be a good thing to test and find out what versions of windows it plays nice with.

I woke up this morning remembering why the truncation.

The edge cases of custom file naming are legion. After this many decades of regular file usage, I don't understand why there isn't a x-plat metadata standard with quick access. As far as I know, no major vendor does this and it should be all of them. Why has every vendor made us put metadata only into this limited name. This isn't exactly a new problem. /rant

Yes, FileManager.FileUtility.GetValidFilename ultimately enforces windows limit safety (another fun gotcha if we go x-platform). However, the audible id is a special field which needs to be preserved. If it's at the end of a really long filename, then it gets trunc'd, then it's gone forever. Something like this would ensure that it's at least in a directory or filename and would have to happen for both regular titles and chapterized titles

    dirName = applyTemplate(dirTemplate, book) // no truncation
    fileName = applyTemplate(fileNameTemplate, book) // no truncation
    finalName = Path.Combine(dirName, fileName)

    idField = $"[{TemplateFields.ID}]"
    if ( ( dirTemplate.Contains(idField ) || fileNameTemplate.Contains(idField) )
        && !finalName.Contains(book.ASIN) )
    {
        trunc =50
        dirName = applyTemplate(dirTemplate, book, trunc)
        fileName = applyTemplate(fileNameTemplate, book, trunc)
        finalName = Path.Combine(dirName, fileName)
    }

and of course the real applyTemplate method (whose name I can't remember) would have to make trunc an optional `int?` and only apply it if `trunc.HasValue`

**Comment by Mbucari** [comment]
> > `\\?\`
> 
> test and find out what versions of windows it plays nice with.

From what I can find, it works a far back as windows XP. I didn't bother trying to find evidence of it working on earlier OSs. I've made a [LongPath](https://github.com/Mbucari/Libation/blob/ae6c2afb3017e346b361ec50e0470dcb8f742b94/Source/FileManager/LongPath.cs) class which is implicitly convertible back and forth from string and converts string paths to extended paths.  This works on local files and network files. This will work on NTFS and FAT32, but many apps (including Windows Flie Explorer) don't play nicely with long paths.  I've tested windows 10, and I can navigate to a long path on the file system and interact with the files inside, but I can't create a long path on the fs using the File Explorer gui.

> the audible id is a special field which needs to be preserved. If it's at the end of a really long filename, then it gets trunc'd, then it's gone forever.

I reworked FileNamingTemplate.GetFilePath ([which you can see here](https://github.com/rmcrackan/Libation/compare/master...Mbucari:master)). How my version works is:

- Split the path into directory and file, and only work with the file (If the path is a directory, the "file" the the bottom directory"
- Build a list of filename parts, replacing template parameters along the way
- While the total filename length is greater than the max filename length, find the longest filename part and remove 1 character from the end of it.

In theory, this approach could still end up truncating the id. Asins are 9 chars long, the max filename length is 255 chars, and the multipart chapter template supports 16 parameters.  In theory, if
- all parameters were used, and
- all parameters are >= 9 chars, and
- the text between all parameters is >= 9 chars

then there would be 33 parts each >= 9 chars, totaling >= 297 chars.  The algo would go through each part, reducing the longest part by 1 until the sum of all parts is 255. That would necessitate removing an average of >1.2 chars from all parts, which would  guarantee the id is truncated.

In practice, few if any people will use ALL parameters, and even if they do, I'd bet nobody will have >=9 chars of text between all parameters.  This approach really should be a fix for everyone, but if you wanted to be extra cautious, I could also exempt the \<id\> parameter from truncation.

**Comment by rmcrackan** [comment]
> I didn't bother trying to find evidence of it working on earlier OSs

Yeah, no point. I'm currently only compiling Libation for win7 x64 and greater

> In theory, this approach could still end up truncating the id

I don't care if it gets trunc'd when someone does something extreme. I want to ensure that one book with a stupidly long title doesn't hurt us. Regarding your math, it'd be dumb yet valid for someone to use a parameter multiple times: `[id] - [title] - [title] - [title]` or to put lots of text in: `my prefix manifesto [title]`

Normally I haven't worried about being cross platform because that was so distant of a possibility. As it looks more feasible, we have to consider that here. Hardcoding `path = path.Replace('/', '\\');` will need to use things like `Path.DirectorySeparatorChar` and `Path.AltDirectorySeparatorChar` instead. I don't think the windows-specific `DllImport` marshalling will actually happen until it's called (not at compile) but I won't swear to it.

**Comment by rmcrackan** [comment]
Added to [version 8.0.3](https://github.com/rmcrackan/Libation/releases/latest)
